llvm/lib/Target/BPF/BPFCheckUnreachableIR.cpp

```diff
@@ -0,0 +1,112 @@
+//===--------- BPFCheckUnreachableIR.cpp - Issue Unreachable Error --------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// Check 'unreachable' IRs and issue proper errors.
+//
+//===----------------------------------------------------------------------===//
+
+#include "BPF.h"
+#include "llvm/IR/Constants.h"
+#include "llvm/IR/DiagnosticInfo.h"
+#include "llvm/IR/Instruction.h"
+#include "llvm/IR/Module.h"
+#include "llvm/IR/Type.h"
+#include "llvm/IR/Value.h"
+#include "llvm/Pass.h"
+
+#define DEBUG_TYPE "bpf-check-unreachable-ir"
+
+using namespace llvm;
+
+static cl::opt<bool>
+    DisableCheckUnreachableIR("bpf-disable-check-unreachable-ir", cl::Hidden,
+                              cl::desc("BPF: Disable Checking Unreachable IR"),
+                              cl::init(false));
+
+namespace {
+
+class BPFCheckUnreachableIR final : public ModulePass {
+  bool runOnModule(Module &F) override;
+
+public:
+  static char ID;
+  BPFCheckUnreachableIR() : ModulePass(ID) {}
+
+private:
+  void BPFCheckUnreachableIRImpl(Function &F);
+  void BPFCheckInst(Function &F, BasicBlock &BB, Instruction &I);
+  void HandleUnreachableInsn(Function &F, BasicBlock &BB, Instruction &I);
+};
+} // End anonymous namespace
+
+char BPFCheckUnreachableIR::ID = 0;
+INITIALIZE_PASS(BPFCheckUnreachableIR, DEBUG_TYPE, "BPF Check Unreachable IRs",
+                false, false)
+
+ModulePass *llvm::createBPFCheckUnreachableIR() {
+  return new BPFCheckUnreachableIR();
+}
+
+void BPFCheckUnreachableIR::HandleUnreachableInsn(Function &F, BasicBlock &BB,
+                                                  Instruction &I) {
+  // LLVM may create a switch statement with default to a 'unreachable' basic
+  // block. Do not warn for such cases.
+  unsigned NumNoSwitches = 0, NumSwitches = 0;
+  for (BasicBlock *Pred : predecessors(&BB)) {
+    const Instruction *Term = Pred->getTerminator();
+    if (Term && Term->getOpcode() == Instruction::Switch) {
+      NumSwitches++;
+      continue;
+    }
+    NumNoSwitches++;
+  }
+  if (NumSwitches > 0 && NumNoSwitches == 0)
+    return;
+
+  // If the previous insn is no return, do not warn for such cases.
+  // One example is __bpf_unreachable from libbpf bpf_headers.h.
+  Instruction *PrevI = I.getPrevNonDebugInstruction();
+  if (PrevI) {
+    auto *CI = dyn_cast<CallInst>(PrevI);
+    if (CI && CI->doesNotReturn())
+      return;
+  }
+
+  F.getContext().diagnose(
+      DiagnosticInfoGeneric(Twine("unreachable in func ")
```

Comment:
I feel that reporting file:line would still be useful in some cases.
If the error could say that in function FOO from line X to line Y that code was deleted as unreachable
that would help users to address the problem.

Reply:
Okay, if debuginfo (-g) is available, the error message will be something like
```
error: in func repro from line 155 to the end of func that code was deleted as unreachable,
          due to uninitialized variable? try -Wuninitialized?
```

Without debuginfo, the error message will be like
```
error: in func repro that code was deleted as unreachable,
          due to uninitialized variable? try -Wuninitialized?
```

Reply:
> error: in func repro from line 155 to the end of func that code was deleted as unreachable,
>           due to uninitialized variable? try -Wuninitialized?

this looks better, maybe add '' around function name like we do elsewhere and don't shorten "function" to "func" ?

Reply:
Sure. Will do.